src/content/en/updates/2018/06/audio-worklet-design-pattern.md

```diff
@@ -0,0 +1,414 @@
+callback so it must not be synchronously blocked. In the worst case scenario the
+audio stream might suffer from duplicate or drop out but it will eventually
+recover when the rendering performance is stabilized.
+
+#### Setting Up and Running
+
+As shown in the diagram above, this design has several components to arrange:
+DedicatedWorkerGlobalScope (DWGS), AudioWorkletGlobalScope (AWGS),
+SharedArrayBuffer and the main thread. The following steps describe what should
+happen in the initialization phase.
+
+##### Initialization
+
+  1. [Main] AudioWorkletNode constructor gets called.
+    1. Create Worker.
+    1. The associated AudioWorkletProcessor will be created.
+  1. [DWGS] Worker creates 2 SharedArrayBuffers. (one for shared states and the
+    other for audio data)
+  1. [DWGS] Worker sends SharedArrayBuffer references to AudioWorkletNode.
+  1. [Main] AudioWorkletNode sends SharedArrayBuffer references to
+    AudioWorkletProcessor.
+  1. [AWGS] AudioWorkletProcessor notifies AudioWorkletNode that the setup is
+    completed.
+
+Once the initialization is completed, `AudioWorkletProcessor.process()` starts
+to get called. The following is what should happen in each iteration of the
+rendering loop.
+
+##### Rendering Loop
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-7.png">
+  <figcaption>
+    Multi-threaded rendering with SharedArrayBuffers
+  </figcaption>
+</figure>
+
+  1. [AWGS] `AudioWorkletProcessor.process(inputs, outputs)` gets called for
+    every render quantum.
+    1. `inputs` will be pushed into **Input SAB**.
+    1. `outputs` will be filled by consuming audio data in **Output SAB**.
+    1. Updates **States SAB** with new buffer indexes accordingly.
+    1. If **Output SAB** gets close to underflow threshold, Wake Worker to
+      render more audio data.
+  1. [DWGS] Worker waits (sleeps) for the wake signal from
+    `AudioWorkletProcessor.process()`. When it wakes up:
+    1. Fetches buffer indexes from **States SAB**.
+    1. Run the process function with data from **Input SAB** to fill
+      **Output SAB**.
+    1. Updates **States SAB** with buffer indexes accordingly.
+    1. Goes to sleep and wait for the next signal.
+
+The example code can be found here, but note that the SharedArrayBuffer
+experimental flag must be enabled for this demo to work. The code was written
+with pure JS code for simplicity, but it can be replaced with WebAssembly code
+if needed. Such case should be handled with extra care by wrapping memory
+management with
+[HeapAudioBuffer](https://github.com/GoogleChromeLabs/web-audio-samples/blob/gh-pages/audio-worklet/design-pattern/lib/wa2.js#L34)
+class.
+
+## Conclusion
+
+The ultimate goal of the AudioWorklet is to make the Web Audio API truly
+"extensible". The multi-year effort went to its design to make it possible to
+implement the rest of Web Audio API with the AudioWorklet. In turn, now we have
```

Comment:
I don't think that was the actual goal.  The goal was an extensible node that looks like a built-in AudioNode. I don't think we really spent much time thinking if it could implement the rest of the API.

Reply:
Hmm. I beg differ. The extensibility was the most important design principle of AW. With that principle, implementing the rest of API should be possible. Even if that's not exactly where we are, I am certain that we're heading that direction.

Reply:
Oh, sorry, I think I was not clear.  The extensibility was the most important goal.  But not the part about being able to implement the existing nodes in WebAudio, but we certainly did want to make it possible, but we didn't think too much about whether it was actually possible.  But being able to do so is a great testament that we got the design right (which remains to be seen still since no one has actually tried).